lib/instrument.ex

```diff
@@ -0,0 +1,390 @@
+  end
+
+  @doc """
+  Enocdes a call given as tuple `{m, f, a}` as Elixir values into an
+  abstract erlang form
+  """
+  def encode_call({m, f, a}) do
+    line = 0 # [generated: true]
+    {:call, line,
+      {:remote, line,
+        {:atom, line, m},
+        {:atom, line, f}},
+      Enum.map(a, &encode_value/1)}
+  end
+  def encode_call(m, f, a) when is_atom(m) and is_atom(f) and is_list(a),
+    do: encode_call({m, f, a})
+
+  @doc "Encodes a value"
+  def encode_value(nil), do: {:nil, [generated: true]}
+  def encode_value(value) when is_atom(value), do: {:atom, [generated: true], value}
+  def encode_value(value) when is_integer(value), do: {:integer, [generated: true], value}
+  def encode_value(value) when is_float(value), do: {:float, [generated: true], value}
+  def encode_value(value) when is_binary(value), do:
+    {:bin, [generated: true],
+      [{:bin_element, [generated: true], {:string, [generated: true], String.to_charlist(value)},
+        :default, :default}]}
+  def encode_value([]), do: encode_value(nil)
+  def encode_value(l) when is_list(l) do
+    l
+    |> Enum.reverse()
+    |> Enum.reduce({nil, 0}, fn v, acc -> {:cons, 0, encode_value(v), acc} end)
+  end
+  def encode_value(t) when is_tuple(t) do
+    {:tuple, [generated: true],
+      t
+      |> Tuple.to_list()
+      |> Enum.map(&encode_value/1)}
+  end
+  def encode_value(_unknown), do: throw ArgumentError
+
+  @doc "Encodes a call to `:erlang.yield()"
+  def call_yield do
+    encode_call({:erlang, :yield, []})
+  end
+
+  # Generate the matcher for interestng functions by a macro
+  # This list is taken from the opensource version of Pulse
+  # (https://github.com/prowessproject/pulse-time/blob/master/src/pulse_time_scheduler.erl)
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :lookup}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :lookup_element}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :update_element}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :insert}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :insert_new}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :delete_object}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :delete}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :delete_all_objects}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :select_delete}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :match_delete}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :match_object}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :member}), do: true
+  def instrumentable_function({:atom, _, :ets}, {:atom, _, :new}), do: true
+
+  def instrumentable_function({:atom, _, :gen_server}, {:atom, _, :start_link}), do: true
```

Comment:
Note that this does not cover `GenServer`. We might be able to instrument `:gen` though, which is used within Elixir's own `GenServer` implementation.